_source/_posts/2019-12-12-tutorial-kotlin-beginners-guide.md

```diff
@@ -0,0 +1,491 @@
+
+**Invariant array**: In Java, arrays are covariant, and array `Integer[]` is a subtype of `Number[]`, so the compiler allows assigning a `Double` to a `Number[]` reference, but the program might raise an exception at runtime if the instance is `Integer[]`. Kotlin does not let you assign `Array<String>` to `Array<Any>`, preventing runtime failures. `Any` is the root of the Kotlin class hierarchy.
+
+```java
+// Compiles but will raise java.lang.ArrayStoreException
+Integer[] intArray = new Integer[1];
+Number[] numberArray = intArray;
+numberArray = intArray;
+numberArray[0] = 1.0;
+```
+
+In Kotlin:
+
+```kotlin
+var intArray = arrayOf(1)
+var numArray = arrayOf(1.0)
+numArray = intArray // Compiler error Type mismatch
+```
+
+`Any` is the root of the Kotlin class hierarchy, but it is not equivalent to `Object`. However, `java.lang.Object` is mapped to `kotlin.Any!`, and the `!` notation means it can be `Any` or `Any?` (nullable or not). As `Object`, other Java types are not used "as is", but mapped to Kotlin types. A complete list of mappings is available in the [Kotlin Reference](https://kotlinlang.org/docs/reference/java-interop.html#mapped-types).
+
+**No raw types**: There are no raw types in Kotlin, as generics are different. Java raw types, for example `List`, is converted into star projections `List<*>!`, which are similar to raw types, but provide runtime safety. How? The compiler will not allow write operations if the type argument is unknown, as it might cause a cast exception when reading. In Java, you can add any object to a raw List, but in Kotlin adding to a start projected list won't compile
+
+**Use-site variance**: Java's type system uses bounded wildcards to increase API flexibility, as generic types are invariant (as opposed to Java arrays, which are covariant), meaning `List<String>` is not a subtype of `List<Object>`, but can be assigned to `List<? extends Object>` type. In Kotlin, instead of bounded wildcards, use-site variance allows to restrict the generic type in the place it is used, with a simpler syntax like `Array<out Any>`, equivalent to Java's `Array<? extends Object>`. The compiler verifies the parameter type is only returned from a method of the instance, but not consumed, to avoid runtime exceptions, for example, if attempting to write a `String` to `Array<Int>`. 
+
+```kotlin
+package com.okta.developer
+
+fun main() {
+
+    val a = arrayOf("uno", "dos", "tres")
+    readMethod(a)
+}
+
+fun readMethod(a: Array<out Any>) {
+    println(a[0])
+    a.set(0, 1) // Out-projected type prohibits the use of fun set
+}
+```
+
+The example above does not compile in Kotlin, because the `readMethod` receives the out-projected array (only read operations allowed) and it is calling the write operation `set`.
```

Comment:
```suggestion
The example above does not compile in Kotlin, because the `readMethod()` receives the out-projected array (only read operations allowed) and it is calling the write operation `set()`.
```